src/components/CPU/BreakpointPanel.js

```diff
@@ -0,0 +1,209 @@
+		if (!context.gameStatus.started) {
+			return;
+		}
+		context.ppsspp.send({
+			event: 'cpu.breakpoint.list',
+		}).then(({ breakpoints }) => {
+			setCpuBreakpoints(breakpoints);
+		}, () => {
+			setCpuBreakpoints([]);
+		});
+	}, [context.ppsspp, context.gameStatus.started]);
+
+	const fetchMemoryBreakpoints = useCallback(() => {
+		if (!context.gameStatus.started) {
+			return;
+		}
+		context.ppsspp.send({
+			event: 'memory.breakpoint.list',
+		}).then(({ breakpoints }) => {
+			setMemoryBreakpoints(breakpoints);
+		}, () => {
+			setMemoryBreakpoints([]);
+		});
+	}, [context.ppsspp, context.gameStatus.started]);
+
+	const fetchBreakpoints = useCallback(() => {
+		fetchCpuBreakpoints();
+		fetchMemoryBreakpoints();
+	}, [fetchCpuBreakpoints, fetchMemoryBreakpoints]);
+
+	const clearBreakpoints = useCallback(() => {
+		setCpuBreakpoints([]);
+		setMemoryBreakpoints([]);
+		setSelectedRow(-1);
+	}, []);
+
+	useEffect(() => {
+		const interval = setInterval(() => {
+			fetchBreakpoints();
+		}, 3000);
+		return () => clearInterval(interval);
+	}, [fetchBreakpoints]);
+
+	useEffect(() => {
+		const listeners_ = listeners.listen({
+			'game.start': () => fetchBreakpoints(),
+			'game.quit': () => clearBreakpoints(),
+			'connection': () => fetchBreakpoints(),
+			'cpu.stepping': () => fetchBreakpoints(),
+			'cpu.breakpoint.add': () => fetchCpuBreakpoints(),
+			'cpu.breakpoint.update': () => fetchCpuBreakpoints(),
+			'cpu.breakpoint.remove': () => fetchCpuBreakpoints(),
+			'memory.breakpoint.add': () => fetchMemoryBreakpoints(),
+			'memory.breakpoint.update': () => fetchMemoryBreakpoints(),
+			'memory.breakpoint.remove': () => fetchMemoryBreakpoints(),
+		});
+		return () => listeners.forget(listeners_);
+	}, [fetchBreakpoints, fetchCpuBreakpoints, fetchMemoryBreakpoints, clearBreakpoints]);
+
+	const handleToggleCpuBreakpoint = (breakpoint) => {
+		context.ppsspp.send({
+			event: 'cpu.breakpoint.update',
+			address: breakpoint.address,
+			enabled: !breakpoint.enabled,
+		});
+	};
+
+	const handleToggleMemoryBreakpoint = (breakpoint) => {
+		context.ppsspp.send({
+			event: 'memory.breakpoint.update',
+			address: breakpoint.address,
+			size: breakpoint.size,
+			enabled: !breakpoint.enabled,
+		});
+	};
+
+	const handleToggleBreakpoint = (breakpoint, type) => {
+		if (type === 'memory') {
+			handleToggleMemoryBreakpoint(breakpoint);
+		} else if (type === 'execute') {
+			handleToggleCpuBreakpoint(breakpoint);
+		}
+	};
+
+	const handleCloseBreakpointModal = () => {
+		setEditingBreakpoint(null);
+		setCreatingBreakpoint(false);
+	};
+
+	const onKeyDown = (ev) => {
+		if (hasContextMenu()) {
+			return;
+		}
+
+		if (ev.key === 'ArrowUp' && selectedRow > 0) {
+			setSelectedRow(selectedRow - 1);
+			ev.preventDefault();
+		}
+		if (ev.key === 'ArrowDown' && selectedRow < breakpoints.length - 1) {
+			setSelectedRow(selectedRow + 1);
+			ev.preventDefault();
+		}
+		const selectedBreakpoint = breakpoints[selectedRow];

```

Comment:
We should probably have ArrowRight gotoDisasm it, just so there's a keyboard way to do that other than double clicking.

-[Unknown]